### PR TITLE
Bubbles

### PR DESCRIPTION
**Pull request by julesduroux**

### CODE REVIEW

**Comment by your-mom-noob** [comment]
Tests failed because there is a problem with the database
Solved it with a migration, but that only works locally...

File: engine_modules/corporation/models.py [hunk 1]
```diff
@@ -78,6 +78,8 @@ def retrieve_all(cls):
 		return cls.base_corporations.values()
 
 # Build the dict at startup once and for all
+# This is executed when the file is imported (this is why 'import from' is important) !
+# Isn't there a better way to do this ?
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: dunno, if you have one i'm in :)

File: engine_modules/corporation/models.py [hunk 2]
```diff
@@ -91,11 +93,18 @@ class Meta:
 
 	base_corporation_slug = models.CharField(max_length=20)
 	game = models.ForeignKey(Game)
-	assets = models.SmallIntegerField()
+	# assets, market_assets and assets_modifier are meant to keep track of the MarketBubbles:
+	# - market_assets stands for the total of the assets in the Corporation's markets disregarding bubbles.
+	# - assets_modifier stands for the bonuses and maluses originating from domination on a market, or having a market at 0 assets.
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: 0 asset-s-

Comment (Neamar) [inline engine_modules/corporation/models.py]: also add it can be negative?

File: engine_modules/corporation/models.py [hunk 3]
```diff
@@ -158,8 +189,13 @@ def update(corporation, delta, market=None):
...
 			else:
-				# TODO; implement and test effects with a market name
-				raise NotImplementedError()
+				for m in corporation.markets:
+					if m.name == market:
+						market = m
+						break
+
+				if isinstance(market, str):
+					raise ValidationError("La corporation %s n'est pas présente sur le marché %s, elle ne peut pas être impactée par un effet sur ce dernier" % (self.base_corporation.name, m))
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: in english, i believe it's a backend error only?

File: engine_modules/corporation/models.py [hunk 4]
```diff
@@ -180,32 +216,31 @@ def on_last_effect(self, ladder):
...
+	def increase_market_assets(self, value=1):
 		"""
-		Increase corporation's assets by value
+		Increase corporation's market_assets by value
 		"""
 		self.market_assets += value
 		self.save()
 
-	def decrease_assets(self, value=1):
+	def decrease_market_assets(self, value=1):
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: when is this used?

Comment (Neamar) [inline engine_modules/corporation/models.py]: Why not use update_market_assets with a delta for instance, as update_assets?

File: engine_modules/corporation/models.py (hunk 1 above)
Comment (Neamar) [inline engine_modules/corporation/models.py]: Idea: make it private, only expose it via corporation.get_base_corporation

File: engine_modules/corporation/models.py [hunk 5]
```diff
@@ -91,6 +93,10 @@ class Meta:
 
 	base_corporation_slug = models.CharField(max_length=20)
 	game = models.ForeignKey(Game)
+	# assets, market_assets and assets_modifier are meant to keep track of the MarketBubbles:
+	# - market_assets stands for the total of the assets in the Corporation's markets disregarding bubbles.
+	# - assets_modifier stands for the bonuses and maluses originating from domination on a market, or having a market at 0 assets.
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: asset

File: engine_modules/corporation/models.py [hunk 6]
```diff
@@ -124,16 +129,39 @@ def get_random_market(self):
...
-			return random.choice(common_corporation_markets)
-		else:
+		if len(common_corporation_markets) == 0:
 			raise ValidationError("Corporations %s and %s have no common market" % (self.base_corporation.name, c2.base_corporation.name))
+		else:
+			return common_corporation_markets
+
+	def get_common_market(self, c2):
+		"""
+		Returns the Market object for a common market between the Corporation if there is one
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: is at least once

File: engine_modules/corporation/models.py [hunk 7]
```diff
@@ -158,8 +186,13 @@ def update(corporation, delta, market=None):
...
 				market = corporation.get_random_market()
 			else:
-				# TODO; implement and test effects with a market name
-				raise NotImplementedError()
+				for m in corporation.markets:
+					if m.name == market:
+						market = m
+						break
+
+				if isinstance(market, str):
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: for else

File: engine_modules/corporation/models.py [hunk 8]
```diff
@@ -180,32 +213,50 @@ def on_last_effect(self, ladder):
...
+		Must be used for all modifications on assets_modifier, because it enforces assets = market_assets + asset_modifier
 		"""
-		self.market_assets += value
+		self.assets_modifier = value
+		self.assets = self.market_assets + self.assets_modifier
+		self.save()
+
+	def set_market_assets(self, value=0):
+		"""
+		This is here to replace the assignments 'c.market_assets = xxx', because they do not enforce assets = market_assets + asset_modifier
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: Add "only for tests"

File: engine_modules/corporation/models.py [hunk 9]
```diff
@@ -180,32 +213,50 @@ def on_last_effect(self, ladder):
...
+		"""
+		self.market_assets = value
+		self.assets = self.market_assets + self.assets_modifier
 		self.save()
 
-	def decrease_assets(self, value=1):
+	def increase_market_assets(self, value=1):
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: i think we can remove this function

File: engine_modules/corporation/models.py [hunk 10]
```diff
@@ -180,32 +213,50 @@ def on_last_effect(self, ladder):
...
 		"""
-		Decrease corporation's assets by value
+		Increase corporation's market_assets by value
 		"""
+		self.market_assets += value
+		self.assets = self.market_assets + self.assets_modifier
+		self.save()
 
+	def decrease_market_assets(self, value=1):
```
Comment (Neamar) [inline engine_modules/corporation/models.py]: we can remove this one too

File: engine_modules/corporation/models.py (hunk 8 above)
Comment (Neamar) [inline engine_modules/corporation/models.py]: or move it to engine/testcases?

**Pull request merged**